libbpfgo/libbpfgo.go

```diff
@@ -260,14 +278,25 @@ func bumpMemlockRlimit() error {
 	return nil
 }
 
+func errnoError(ptr unsafe.Pointer, errFmt string, args... interface{}) error {
+	negErrno := C.PTR_ERR(ptr)
+	errno := syscall.Errno(-int64(negErrno))
+	if errno == 0 {
+		return fmt.Errorf(errFmt, args...)
+	} else {
+		args := append(args, errno.Error())
+		return fmt.Errorf(errFmt + ": %v", args...)
+	}
+}
+
 func NewModuleFromFile(bpfObjFile string) (*Module, error) {
 	C.set_print_fn()
 	bumpMemlockRlimit()
 	cs := C.CString(bpfObjFile)
 	obj := C.bpf_object__open(cs)
 	C.free(unsafe.Pointer(cs))
-	if obj == nil {
-		return nil, fmt.Errorf("failed to open BPF object %s", bpfObjFile)
+	if C.IS_ERR_OR_NULL(unsafe.Pointer(obj)) {
+		return nil, errnoError(unsafe.Pointer(obj), "failed to open BPF object %s", bpfObjFile)
```

Comment:
Maybe this is a good time to add a unit test for this function in order to avoid such scenarios in the future.

Reply:
ditto elsewhere where the check is done

Reply:
I think it would be beneficial to use a unit testing framework (builtin or `check`) instead of standalone selftests.
Should that be a separate issue/PR though?

Reply:
A simple unit test would suffice IMO. I usually like to write the test for the change of code along with it.

Reply:
I agree, I'm saying that libbpfgo as a whole is missing unit tests. It would be beneficial for write the tests that cover the whole lib.

But I'll kick things off by adding at least a test for a one or two of these functions.

Reply:
I added the tests but nothing is running them yet. I used `//go:generate` instead of a `Makefile` (can change it) so the CI would need to run `go generate` prior to running `go test`. PTAL and let me know how to best invoke them in CI.

The tests also require 5.8+ kernel because of `AttachLSM` although I can put it behind a flag or something.

Reply:
/cc @simar7

Reply:
Thanks I will take a look at it shortly and get back to you.

Reply:
> I added the tests but nothing is running them yet. I used `//go:generate` instead of a `Makefile` (can change it) so the CI would need to run `go generate` prior to running `go test`. PTAL and let me know how to best invoke them in CI.
> 
> The tests also require 5.8+ kernel because of `AttachLSM` although I can put it behind a flag or something.

Thanks this looks wonderful @eyakubovich! Looks like it just needs to be told where to find the right headers:

```
vagrant@ubuntu-groovy:/usr/vagrant/repos/tracee/libbpfgo$ go generate
test/test.bpf.c:3:10: fatal error: 'bpf/bpf_helpers.h' file not found
#include <bpf/bpf_helpers.h>
         ^~~~~~~~~~~~~~~~~~~
1 error generated.
libbpfgo_test.go:2: running "clang": exit status 1
```

As you mentioned it would be better to add them to run via the Makefile as some of the pre-requisite headers will be set in it. We currently do this for running perfbuffer and ringbuffer tests and you can find an example here: https://github.com/aquasecurity/tracee/blob/77cf435059442695b6b604630406ddc8614820e8/libbpfgo/selftest/perfbuffers/Makefile#L4-L7

Reply:
@simar7 PTAL. The tests fail on CI (as expected, kernel version) so I didn't add them to the GH actions. Can you check if they run for you? I'm a bit surprised that `#include <bpf/bpf_helpers.h>` failed -- if libbpf is installed, it should be in `/usr/include` and should work. If it's installed elsewhere, I guess the user will need to specify.

Reply:
Once we sort the running of the tests locally, we are good to merge this in. Appreciate your patience.